torch/_dynamo/variables/tensor.py

```diff
@@ -681,13 +684,49 @@ def redistribute_fn_with_prim_types(x):
                 mutable_local=variables.base.MutableLocal(),
                 **options,
             )
+
             if not self.source:
                 # Intermediary
-                unimplemented("Intermediary tensors with registered hooks - NYI")
-            else:
-                assert (
-                    fn_var.source
-                ), "Unreachable - See unimplemented for lambdas above"
+                src = fn_var.source
+                if (
+                    not src
+                    and isinstance(fn_var, variables.functions.FunctoolsPartialVariable)
+                    and fn_var.func.source
+                ):
+                    src = fn_var.func.source
+
+                if src:
+                    tx.output.guards.add(src.make_guard(GuardBuilder.ID_MATCH))
```

Comment:
So, in the simplified case, the functools partial is on a function and a var that both have sources (i.e., I don't have to do any reconstruction before calling the register hook.) In that case, this set of guards seems insufficient; I also have to guard on the functools arguments, no?

Reply:
There isn't really any checking to see if we're in the simplified case and that also makes me nervous

Reply:
Let's add a propagate here, that should be enough, the fn_var should already have those.